### ISSUE TITLE
Iteration with nested Unicorn component doesn't work

### ISSUE THREAD

**Issue by gschurck** [adamghill/django-unicorn#526]
On my website I want to iterate on a queryset with a for loop, but when I pass each object to an Unicorn component, the object displayed is always the same of the list, but multiple times. And it corresponds with the size of the list...

When I display each object data during iteration in the parent component it works, but not with a nested component.

Is it normal ? How to fix this ?
Thanks.

![image](https://user-images.githubusercontent.com/21091232/225679073-bfbf64cc-198a-4b79-b53e-0249541cfac4.png)

---

Parent Unicorn component: `home.html`
```html
...
|{{ memes_results }}|
<div id="loading-images">
    <div class="">
        {% for meme_temp in memes_results %}
           Parent:  ( {{ meme_temp.title }} )
            <br>
            <div class="column is-one-quarter-desktop is-half-touch">
                {% unicorn 'meme_card' meme=meme_temp title=meme_temp.title %}
            </div>
        {% endfor %}
    </div>
</div>
...
```
<br>

Nested Unicorn component`meme_card.html`

```html
{% load unicorn %}
<div class="card">
    Nested component: {{ title }}
</div>
```

**Comment by adamghill** [comment]
It's hard to tell what might be going on here from this explanation. If you make a repo that reproduces the behavior I can dig into it. Thanks!

**Comment by gschurck** [comment]
Here it is : 
https://github.com/gschurck/django-unicorn-demo/tree/main

I expect the unicorn component to be callable and nestable in a loop in the same way as classic Django templates.

So I don't know if it is a bug or a normal behavior, but I think it is a huge use-case and it should be possible.

![image](https://user-images.githubusercontent.com/21091232/226144650-80552d6b-611d-4690-aaad-5321645be656.png)

**Comment by adamghill** [comment]
Thanks for the repo to replicate your issue. 

I made a [PR](https://github.com/gschurck/django-unicorn-demo/pull/1) that is one way to alleviate your current your problem.

The reason this is happening is because `Unicorn` aggressively caches components so they do not need to be re-constructed on every request. The cache key is based on a few kwargs for the component template tag, e.g. `name`, `id`, and `key`. This is sort of [implied by the docs](https://github.com/adamghill/django-unicorn/blob/de4c66376ea5f5f21f51ef5e4ae5b84dede3e378/docs/source/components.md?plain=1#L30-L32), but isn't super clear.

As an enhancement I could base the cache key on _all_ of the `kwargs` passed into the component template tag. That would mean that each of your nested components would be cached separately because each `text` kwarg is unique. Or I could update the docs to be more clear? What do you think?

**Comment by gschurck** [comment]
Thanks for your help !

> This is sort of [implied by the docs](https://github.com/adamghill/django-unicorn/blob/de4c66376ea5f5f21f51ef5e4ae5b84dede3e378/docs/source/components.md?plain=1#L30-L32), but isn't super clear.

Oh I didn't understand there was a component caching system, the documentation doesn't say for what the keys are used for.

> As an enhancement I could base the cache key on all of the kwargs passed into the component template tag. That would mean that each of your nested components would be cached separately because each text kwarg is unique. Or I could update the docs to be more clear? What do you think?

What is the purpose of this cache ? How is it used ? Is it very needed by default everywhere ?

Personally, I expect a for loop with a nested component to work out of the box without adding any unique key identifier, as it is the case for the vast majority of other component-based templating system. It feels like a bit counter-intuitive for me.

For me the best solution is : 
- Either make component caching optional if it is not really useful everywhere
- Or if it is useful, using all the kwargs as caching key could be a seamless solution, if there are no issues with this. Just maybe add the possibility to override the key if needed.

And the documentation will need a clear section on component caching, for sure !

**Comment by adamghill** [comment]
>using all the kwargs as caching key could be a seamless solution

Yeah, I think this is the way to go. I have an approach in mind -- I'll update this issue with what I figure out. 👍

**Comment by adamghill** [comment]
I think this should be fixed by 0.58.0 of `Unicorn` which drastically improves child components so I'm going to close it. If it is still a problem feel free to re-open this. Thank you!
